### DISCUSSION TITLE
Accessing q-page

### DISCUSSION THREAD

**Discussion by shakamoushie** [zauberzeug/nicegui#340] [Q&A] [answered]
How do I write content on the q-page?

Problem:
1. I have a main page, which contains a tree in the left drawer,
2. On clicking one of the tree options, I go into the tree callback
3. In the tree callback, based on some conditions, I want to output some new text in the main content area of the page (q-page), after clearing previous contents (if any).

How do I do this?

Thanks in advance...

**Comment by falkoschindler** [comment, marked as answer]
To access the `q-page` element, you need to use a page function with a `ui.page` decorator. This function accepts a `Client` object which has a `content` field:

```py
from nicegui import Client, ui

@ui.page('/')
def page(client: Client):
    def add_content():
        with client.content:
            ui.label('Content')
    with ui.left_drawer().classes('bg-gray-100'):
        ui.button('Add content', on_click=add_content)

ui.run()
```

---

Alternatively, you can simply define some arbitrary container that is accessible from within the event handler:

```py
from nicegui import ui

container = ui.column()

def add_content():
    with container:
        ui.label('Content')

with ui.left_drawer().classes('bg-gray-100'):
    ui.button('Add content', on_click=add_content)

ui.run()
```

This solution does not strictly add elements to `q-page` but to a child element, but in most cases that's ok.

**Reply by shakamoushie** [reply]
Thanks @falkoschindler for your immediate response, will try out the same and get back to you.

Cheers

**Reply by shakamoushie** [reply]
Thanks @falkoschindler and @rodja , your code works (...obviously... ;-)), but I want to ask if I can clear the main section / q-page without using a button

```py
from nicegui import Client, ui

@ui.page('/')
def page(client: Client):
    def ExecuteCalc(vval):
        # ui.timer(interval=1.0, callback=lambda: client.content.clear)
        if vval == 'BSA':
            ui.label("BSA calc starting...")
        else:
            ui.label("Some other calc starting...")

    def set_wch_calc(vval):
        with client.content:
            # ui.timer(interval=1.0, callback=lambda: client.content.clear)
            ui.html(f'<h1>You chose {vval}</h1>')
            ui.button('Go to Calc', on_click=lambda _: ExecuteCalc(vval))
            
    with ui.left_drawer().classes('bg-gray-100'):
        ui.tree([
                {'id': 'Drug Dose', 'children': [{'id': 'BSA'}]},
                {'id': 'Oncology', 'children': [{'id': 'Recist S'}, {'id': 'Recist M'}, {'id': 'Recist A'}]},
                {'id': 'Gastroenterology', 'children': [{'id': 'Mayo Score'}, {'id': 'Harvey Bradshaw'}, {'id': 'SCCAI'}, {'id': 'SESCD'}]},
            ], label_key='id', on_select=lambda e: set_wch_calc(e.value))
        
        ui.button('Clear', on_click=client.content.clear)

ui.run()
```

In the code above, the Clear button clears the q-page, but I dont want to use a button to do this - I want it to happen automatically. So, I tried using the ui.timer wherever I wanted to clear the q-page, but that does not work.

I am attempting to do the following:
1. Choose a calculator by clicking on an option of the ui.tree in the left drawer
2. Display a little description of the calculator clicked (with ui.html) on the q-page, along with a button to further go to a function (ExecuteCalc) that actually does the calculation
3. Wherever I put ui.timer, I want to issue a command (without user interaction) to clear the q-page.

I would be happy and grateful if you can show me how.

**Reply by rodja** [reply]
It works as expected with small modifications. First, you should use `once=True` for the timer, so it does not execute every second but only one time. Also you need to remove the lambda or put `()` behind the clear method.

**Reply by shakamoushie** [reply]
Awesome! Thanks @rodja  for your help.

Cheers

**Comment by rodja** [comment]
You can also use the `index_client` when you do not want to use page-builder functions. See the discussion https://github.com/zauberzeug/nicegui/discussions/310#discussioncomment-4857986.

**Reply by falkoschindler** [reply]
Oh, right! I forgot about this possibility.

**Comment by bajirama** [comment]
Thanks for this discussion ,  really helpful .  One question how to refresh the client page instead of adding the content ?

**Reply by falkoschindler** [reply]
Hi @bajirama, this discussion is pretty old, possibly outdated, and - without reading the whole discussion - I'm not sure what you mean. Can you, please, open a new Q&A discussion and explain in a bit more detail what you're trying to achieve? Thanks!
